src/main/java/com/wanderersoftherift/wotr/entity/portal/RiftEntrance.java

```diff
@@ -7,85 +7,71 @@
-                    ResourceKey.codec(Registries.DIMENSION)
-                            .fieldOf("target_dimension")
-                            .forGetter(RiftEntranceAttachment::target),
-                    Codec.BOOL.fieldOf("is_generated").forGetter(RiftEntranceAttachment::generated)
-            ).apply(instance, RiftEntranceAttachment::new)
+                    ItemStack.CODEC.fieldOf("key_item").forGetter(RiftEntrance::keyItem),
+                    ResourceKey.codec(Registries.DIMENSION).fieldOf("target_dimension").forGetter(RiftEntrance::target),
+                    Codec.BOOL.fieldOf("is_generated").forGetter(RiftEntrance::generated)
+            ).apply(instance, RiftEntrance::new)
     );
-    private ItemStack keyItem;
-    private ResourceKey<Level> target;
-    private boolean generated;
 
-    public RiftEntranceAttachment(ItemStack keyItem, ResourceKey<Level> target, boolean generated) {
-        this.keyItem = keyItem;
-        this.target = target;
-        this.generated = generated;
-    }
+    public static final StreamCodec<RegistryFriendlyByteBuf, RiftEntrance> STREAM_CODEC = StreamCodec.composite(
+            ItemStack.OPTIONAL_STREAM_CODEC, RiftEntrance::keyItem, ResourceKey.streamCodec(Registries.DIMENSION),
+            RiftEntrance::target, ByteBufCodecs.BOOL, RiftEntrance::generated, RiftEntrance::new);
 
-    public RiftEntranceAttachment() {
-        this(ItemStack.EMPTY,
+    public static RiftEntrance create() {
+        return new RiftEntrance(ItemStack.EMPTY,
                 ResourceKey.create(Registries.DIMENSION, WanderersOfTheRift.id("rift_" + UUID.randomUUID())), false);
     }
 
-    public ItemStack keyItem() {
-        return keyItem;
-    }
-
-    public ResourceKey<Level> target() {
-        return target;
-    }
-
-    public void keyItem(ItemStack keyItem) {
-        this.keyItem = keyItem;
+    public static RiftEntrance loadRiftEntrance(CompoundTag tag, RegistryAccess registries) {
+        var dataResult = RiftEntrance.CODEC.decode(registries.createSerializationContext(NbtOps.INSTANCE),
+                tag.get("entrance_data"));
+        return dataResult.getOrThrow().getFirst();
     }
 
-    public void target(ResourceKey<Level> target) {
-        this.target = target;
+    public void saveRiftEntrance(CompoundTag tag, RegistryAccess registries) {
+        var encodedResult = RiftEntrance.CODEC.encodeStart(registries.createSerializationContext(NbtOps.INSTANCE),
+                this);
+        tag.put("entrance_data", encodedResult.getOrThrow());
     }
 
-    public boolean generated() {
-        return generated;
-    }
-
-    public void generated(boolean generated) {
-        this.generated = generated;
-    }
-
-    public void teleportPlayer(ServerPlayer player, ServerLevel level, Vec3i currentPosition) {
-        var riftId = target;
+    public boolean teleportPlayer(ServerPlayer player, ServerLevel level, Vec3i currentPosition) {
```

Comment:
Tiny thing, but would be good to have a javadoc comment for the meaning of the return value (not obvious it would be whether the rift is generated).